python/robot_fingers/tasks/move_cube.py

```diff
@@ -0,0 +1,165 @@
+        [+1, -1, +1],
+        [+1, +1, -1],
+        [+1, +1, +1],
+    ]
+) * (_CUBE_WIDTH / 2)
+
+
+def get_cube_corner_positions(position, orientation):
+    """Get the positions of the cube's corners with the given pose.
+
+    Args:
+        position: Position (x, y, z) of the cube.
+        orientation: Quaternion (x, y, z, w) with the orientation of the cube.
+
+    Returns:
+        (array, shape=(8, 3)): Positions of the corners of the cube in the
+            given pose.
+    """
+    rotation = Rotation.from_quat(orientation)
+    translation = np.asarray(position)
+
+    return rotation.apply(_cube_corners) + translation
+
+
+def sample_goal(difficulty, current_position=None, current_orientation=None):
+    """Sample a goal pose for the cube.
+
+    Args:
+        difficulty (int):  Difficulty level.  The higher, the more difficult is
+            the goal.  Possible levels are:
+
+            - 1: Goal is on the ground with a random rotation around the
+                  z-axis.
+            - 2: The goal can be above the ground.  Rotation is limited to the
+                  z-axis.
+
+        current_position:  Current (x, y, z)-position of the cube.  Currently
+            unused.
+        current_orientation:  Current orientation of the cube as quaternion
+            (x, y, z, w).  If set, it is considered while sampling the goal
+            orientation.  If not set, it is assumed that the cube is currently
+            aligned with the axes of the world frame.  This is relevant for
+            some difficulty levels to ensure that the goal orientation only
+            differs from the current one by what is specified for that
+            difficulty level (e.g. only rotation around z-axes for level 1).
+
+    Returns:
+        (tuple): A tuple with the goal position (x, y, z) and the goal
+        orientation quaternion (x, y, z, w) relative to the world frame.
+    """
+    # difficulty -1 is for initialization
+
+    # TODO Should we add a minimum distance to current_position to avoid goals
+    # that are too easy?
+
+    # sample uniform position in circle (https://stackoverflow.com/a/50746409)
+    radius = _max_cube_com_distance_to_center * np.sqrt(random.random())
+    theta = random.uniform(0, 2 * np.pi)
+
+    # x,y-position of the cube
+    x = radius * np.cos(theta)
+    y = radius * np.sin(theta)
+
+    if difficulty == -1 or difficulty == 1:
+        # on the ground, random yaw
+        z = _CUBE_WIDTH / 2
+    elif difficulty == 2:
+        # in the air, random yaw
+        z = random.uniform(_min_height, _max_height)
+    else:
+        raise ValueError("Invalid difficulty %d" % difficulty)
+
+    position = np.array((x, y, z))
+
+    # random yaw angle (relative to current_orientation if given)
+    yaw = random.uniform(0, 2 * np.pi)
+
+    orientation = Rotation.from_euler("z", yaw).as_quat()
+    if current_orientation is not None:
+        orientation = orientation * current_orientation
```

Comment:
i guess this does actually proper application of a rotation (i.e. not just elment-wise multiplication)? is this also guaranteed if the current_orientation is simply a numpy array (representing a quaternion)? just to make sure that there is no potential for error here

Reply:
Good point, I missed that.  Yes, the intention is to combine the two rotations but for this `current_orientation` needs to be converted to a `Rotation` object first (this part is not properly tested yet).

Reply:
Fixed.